### ISSUE TITLE
[Feature] Add masked and transparent foreground training modes

### ISSUE THREAD

**Issue by lanxinger** [amirhossein-razlighi/mlx3D#28]
**Is your feature request related to a problem? Please describe.**

Object-centric captures often include a turntable, hands, changing backgrounds, or segmentation mattes that should not contribute to Gaussian Splatting training. mlx3D currently loads COLMAP images as RGB, drops embedded alpha, and computes L1 + SSIM over the full frame. The capture and evaluation paths also have no way to associate a foreground mask with each image.

This can train unwanted background Gaussians, make foreground quality dependent on the capture environment, and prevent clean transparent-object exports. Full-frame evaluation can also hide a foreground improvement behind arbitrary background error.

I searched the current issues for mask, foreground, alpha, transparency, and background support and did not find an existing request.

**Describe the solution you'd like**

Add optional per-image foreground masks with two explicit training semantics:

1. **Masked mode**
   - Treat alpha/mask values as weights for the RGB photometric loss.
   - Black pixels are ignored, white pixels are included, and intermediate values act as soft weights.
   - Normalize L1, SSIM, and other masked reductions by included mask weight so loss scale does not depend on how much of the frame the object occupies.
   - Intersect the ground-truth mask with the valid region used by 2DGS depth-variance and normal-consistency losses.
   - Pixels outside the mask should contribute no loss or gradient.

2. **Transparent mode**
   - Treat alpha/mask values as target foreground opacity.
   - Composite target foreground RGB over the same background used by the renderer.
   - Supervise the differentiable rendered alpha in addition to RGB.
   - Optionally randomize the training background to discourage color leakage around the silhouette and produce a clean transparent result.

The renderer already returns differentiable alpha, so both modes should be possible without changing the rasterizer or Metal kernels.

For external masks, adopting COLMAP's existing naming convention would let one mask set serve training, evaluation, and optional feature extraction:

```text
scene/
  images/abc/frame_001.jpg
  masks/abc/frame_001.jpg.png
  sparse/0/...
```

Requested masks should fail early when files are missing, all-zero, or dimensionally mismatched. External grayscale/RGBA masks would be the most useful initial source; embedded source-image alpha could also be supported explicitly without silently changing current RGB loading behavior.

An illustrative API/CLI could be:

```python
ds = load_colmap(scene, masks_dir="masks", alpha_mode="masked")
cam, image = ds[i]                     # preserve the existing tuple API
trainer.step(cam, image, mask=ds.masks[i])
```

```text
mlx3d-capture photos/ --masks masks/ --alpha-mode masked
mlx3d-capture photos/ --masks masks/ --alpha-mode transparent
mlx3d-eval point_cloud.ply --data scene/ --masks masks/ --alpha-mode masked
```

A small mask collection alongside `ImageCollection` could preserve the existing `ram`, `uint8`, and `disk` cache modes while returning aligned `(H, W, 1)` float masks.

SSIM needs special handling: the current helper reduces SSIM to a scalar internally. Masking that scalar would not isolate background pixels because SSIM uses local windows. A mask-aware implementation should expose a per-pixel SSIM map and define how windows crossing the mask boundary are handled—for example, by excluding windows that overlap ignored pixels or filling ignored pixels identically before computing local statistics.

Training masks and SfM feature masks should be separate controls. A flag such as `--mask-sfm` could optionally:

- pass `--ImageReader.mask_path` to external COLMAP;
- load and resize the paired mask for built-in SfM, then pass it to `SIFT.detectAndCompute`.

This should not happen automatically: background features can help pose recovery for small or low-texture objects even when those pixels should be ignored during splat training.

Evaluation should use the same selected semantics:

- coverage-normalized masked L1 and MSE/PSNR;
- mask-aware SSIM using the same boundary policy as training;
- mask coverage in JSON output;
- for transparent mode, alpha MAE and optionally foreground IoU.

**Describe alternatives you've considered.**

- Pre-compositing every image onto a constant background still trains the chosen background and cannot exclude changing-background errors.
- Cropping images loses camera calibration unless intrinsics are adjusted and is awkward for irregular silhouettes.
- Multiplying the already-reduced scalar SSIM value by a mask does not prevent local SSIM windows from including background pixels.
- Automatically adding a segmentation model would add a large dependency. Accepting paired masks provides the core capability while keeping capture dependencies optional.
- Using only masked RGB loss cannot guarantee a clean transparent silhouette; that is why masked and transparent modes should remain distinct.

**Additional context**

Relevant current paths:

- Image loading drops alpha: [`datasets/images.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/datasets/images.py#L22-L45)
- COLMAP datasets currently carry no masks: [`datasets/colmap.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/datasets/colmap.py#L21-L35)
- Full-frame training loss: [`splatting/trainer.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/splatting/trainer.py#L261-L316)
- SSIM is reduced internally: [`losses/image_metrics.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/losses/image_metrics.py#L29-L85)
- COLMAP feature extraction has no mask argument: [`capture/colmap_wrap.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/capture/colmap_wrap.py#L43-L86)
- Built-in SIFT currently receives `None` as its mask: [`capture/sfm.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/v0.3.0/src/mlx3d/capture/sfm.py#L304-L323)
- COLMAP's documented mask layout: [Mask image regions](https://colmap.github.io/faq.html#mask-image-regions)

Suggested acceptance criteria:

- No-mask behavior remains unchanged.
- Image and mask stay aligned across all cache modes and downscales.
- Requested missing, invalid, or mismatched masks fail with clear diagnostics.
- In masked mode, changing target RGB outside the mask does not affect loss or gradients.
- Masked loss magnitude is independent of foreground coverage.
- In transparent mode, target alpha supervises rendered alpha and background compositing is consistent.
- 2DGS geometry losses respect the selected mask.
- Evaluation uses the same mode and mask semantics as training.
- SfM masks are applied only with an explicit opt-in.
- Dataset, loss, CLI, and capture tests use small generated fixtures and require no downloads.

**Comment by amirhossein-razlighi** [comment]
Thanks for the really thorough writeup, @lanxinger. This is a legitimate gap and you've scoped it well. I went through the current code and everything you flagged is accurate:

- Image loading drops alpha: [`datasets/images.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/datasets/images.py)
- COLMAP datasets currently carry no masks: [`datasets/colmap.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/datasets/colmap.py)
- Full-frame training loss: [`splatting/trainer.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/splatting/trainer.py)
- SSIM is reduced internally: [`losses/image_metrics.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/losses/image_metrics.py)
- COLMAP feature extraction has no mask argument: [`capture/colmap_wrap.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/capture/colmap_wrap.py)
- Built-in SIFT currently receives `None` as its mask: [`capture/sfm.py`](https://github.com/amirhossein-razlighi/mlx3D/blob/main/src/mlx3d/capture/sfm.py)

I also agree with the core distinction you draw: **masked-loss mode** (weight the photometric/SSIM loss by a foreground mask) and **transparent/alpha-supervision mode** (additionally supervise the rendered alpha against the mask so you get a clean cut-out with a transparent background) are genuinely different features with different outputs, and it's right to treat them separately rather than collapsing them.

A quick question: what's your mask source/format: separate grayscale PNGs matched by filename, or alpha embedded in RGBA inputs?

If you're up for implementing any of these I'm glad to review. Otherwise I'll queue it for myself to do soon.

**Comment by lanxinger** [comment]
@amirhossein-razlighi thanks for the quick reply. 

Personally I use separate 8-bit grayscale PNGs over embedded alpha in RGBA but I guess it would be good to support both as I have seen people use alpha also in demo datasets.

If I have some time over the next few days I can take a look at it but have to get more familiar with your library first.
